### PR TITLE
Add recoverable errors

### PR DESCRIPTION
**Pull request by malik672**
fix issue #26 
bump to 0.9.0
introduce recoverable errors

### CODE REVIEW

File: .DS_Store
Comment (shuhuiluo) [inline .DS_Store:1]: You have to `git rm --cached .DS_Store` since it's already added to git indexing.

File: src/.DS_Store
Comment (shuhuiluo) [inline src/.DS_Store:1]: You have to `git rm --cached .DS_Store` since it's already added to git indexing.

**Review by shuhuiluo — CHANGES_REQUESTED** [review]
Will review later.

File: src/.DS_Store
Comment (malik672) [reply src/.DS_Store:1]: ok

File: src/entities/fractions/fraction.rs [hunk 1]
```diff
@@ -253,20 +257,23 @@ mod tests {
...
+        let fraction2 = Fraction::new(4, 12);
+        let expected = Fraction::new(52, 120);
+        let add = fraction1 + fraction2;
+        assert_eq!(add, expected);
+
+        let fraction3 = Fraction::new(1, 5);
+        let fraction4 = Fraction::new(2, 5);
+        let expected2: FractionLike<()> = Fraction::new(3, 5);
+        let add2 = fraction3 + fraction4;
+        assert_eq!(add2, expected2);
```
Comment (shuhuiluo) [inline src/entities/fractions/fraction.rs]: What's wrong with the original test?

File: src/entities/fractions/fraction.rs [hunk 2]
```diff
@@ -297,23 +304,23 @@ mod tests {
...
+        assert!(Fraction::new(1, 10) <= Fraction::new(4, 12));
+        assert!(Fraction::new(1, 3) <= Fraction::new(4, 12));
         assert!(Fraction::new(5, 12) > Fraction::new(4, 12));
     }
 
     #[test]
     fn test_multiply() {
         assert_eq!(
-            Fraction::new(1, 10) * Fraction::new(4, 12),
+            (Fraction::new(1, 10) * Fraction::new(4, 12)),
```
Comment (shuhuiluo) [inline src/entities/fractions/fraction.rs:313]: Adding redundant parenthesis only increases the cost to review.

File: src/entities/fractions/fraction.rs [hunk 3]
```diff
@@ -338,7 +345,7 @@ mod tests {
     fn test_as_faction() {
         let f = Fraction::new(1, 2);
         // returns an equivalent but not the same reference fraction
-        assert_eq!(f.as_fraction(), f);
-        assert_ne!(&f as *const _, &f.as_fraction() as *const _);
+        assert_eq!(f.clone().as_fraction(), f.clone());
+        assert_ne!(&f.clone() as *const _, &f.clone().as_fraction() as *const _);
```
Comment (shuhuiluo) [inline src/entities/fractions/fraction.rs]: There is no need to make clones if the code was working as intended.

File: src/entities/fractions/percent.rs [hunk 4]
```diff
@@ -22,12 +22,14 @@ impl Percent {
     /// Converts the Percent to a string with a specified number of significant digits and rounding strategy
     pub fn to_significant(&self, significant_digits: u8, rounding: Rounding) -> String {
         // Convert the Percent to a simple Fraction, multiply by 100, and then call to_significant on the result
-        (self.as_fraction() * ONE_HUNDRED.clone()).to_significant(significant_digits, rounding)
+        (self.as_fraction() * ONE_HUNDRED.clone())
+            .to_significant(significant_digits, rounding)
+            .unwrap()
```
Comment (shuhuiluo) [inline src/entities/fractions/percent.rs]: The base method returns `Result<String, Error>`. Why is it unwrapped here?

File: src/entities/fractions/percent.rs [hunk 5]
```diff
@@ -22,12 +22,14 @@ impl Percent {
...
     }
 
     /// Converts the Percent to a string with a fixed number of decimal places and rounding strategy
     pub fn to_fixed(&self, decimal_places: u8, rounding: Rounding) -> String {
-        // Convert the Percent to a simple Fraction, multiply by 100, and then call to_fixed on the result
+        // Convert the Percent to a simple Fraction, multiply by 100, and then call to_fixed on the
```
Comment (shuhuiluo) [inline src/entities/fractions/percent.rs]:
```suggestion
        // Convert the Percent to a simple Fraction, multiply by 100, and then call to_fixed on the result
```
Please don't accidentally delete comments.

File: src/entities/fractions/percent.rs [hunk 6]
```diff
@@ -39,47 +41,47 @@ mod tests {
     #[test]
     fn test_add() {
         assert_eq!(
-            Percent::new(1, 100) + Percent::new(2, 100),
+            (Percent::new(1, 100) + Percent::new(2, 100)),
```
Comment (shuhuiluo) [inline src/entities/fractions/percent.rs:46]: These are just unnecessary diffs.

File: src/entities/fractions/price.rs [hunk 7]
```diff
@@ -74,29 +74,32 @@ where
...
-            "TOKEN"
-        );
-        let fraction = self.as_fraction() * other.as_fraction();
-        Price::new(
+    ) -> Result<Price<TBase, TOtherQuote>, Error> {
+        if !self.meta.quote_currency.equals(&other.meta.base_currency) {
+            return Err(Error::NotEqual("the comparison are not equal".to_owned()));
+        }
+
+        let fraction = (self.as_fraction() * other.as_fraction()).clone();
```
Comment (shuhuiluo) [inline src/entities/fractions/price.rs]: A clone is unnecessary.

File: src/entities/fractions/price.rs [hunk 8]
```diff
@@ -140,48 +147,56 @@ mod test {
...
         let price = Price::new(TOKEN0.clone(), TOKEN1.clone(), 1, 54321);
-        assert_eq!(price.to_significant(5, Rounding::RoundDown), "54321");
-        assert!(price.meta.base_currency.equals(&TOKEN0.clone()));
-        assert!(price.meta.quote_currency.equals(&TOKEN1.clone()));
+        assert_eq!(
+            price.to_significant(5, Rounding::RoundDown).unwrap(),
+            "54321"
+        );
+        assert!(price.clone().meta.base_currency.equals(&TOKEN0.clone()));
+        assert!(price.clone().meta.quote_currency.equals(&TOKEN1.clone()));
```
Comment (shuhuiluo) [inline src/entities/fractions/price.rs:155]: `Price` is `Clone` but `&Price` is `Copy`. `fn meta(&self)` takes a reference of `price` which is copied automatically. Please don't make unnecessary clones.
https://doc.rust-lang.org/std/marker/trait.Copy.html

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: src/entities/fractions/fraction.rs (hunk 1 above)
Comment (malik672) [reply src/entities/fractions/fraction.rs]: nothing, the changes I made were to handle result type before, now that's it's no longer result type, will abastract the tests when during the next pr

File: src/entities/fractions/price.rs (hunk 7 above)
Comment (malik672) [reply src/entities/fractions/price.rs]: yep

File: src/entities/fractions/price.rs (hunk 8 above)
Comment (malik672) [reply src/entities/fractions/price.rs:155]: i know though but I probably must have implemented the clones when I was handling something using the result type for the functions, I will clean all the test and the codebase during the next pr, probably today

**Review by shuhuiluo — APPROVED** [review]

File: src/entities/fractions/percent.rs (hunk 6 above)
Comment (malik672) [reply src/entities/fractions/percent.rs:46]: is there a way clippy can discover this

**Pull request merged**